### PR TITLE
Caleb/train tf models

### PR DESCRIPTION
**Pull request by inf0rmatiker**
Train/save TF models

### CODE REVIEW

File: modelservice/worker.py [hunk 1]
```diff
@@ -72,12 +78,118 @@ def register(self):
...
+        info(f"Finished training {count}/{len(request.gis_joins)} models. Returning results...")
+        worker_timer.stop()
+        return WorkerBuildModelsResponse(
             id=request.id,
             hostname=self.hostname,
-            duration_sec=0.0,  # TODO: Capture job profile
-            error_occurred=True,
-            error_msg="Building models currently unimplemented"
+            duration_sec=worker_timer.elapsed,
+            error_occurred=False,
```
Comment (NFabrizio) [inline modelservice/worker.py:190]: Not sure that we need to, but do we want to add some logic for handling errors?

File: modelservice/worker.py [hunk 2]
```diff
@@ -72,12 +78,118 @@ def register(self):
...
+            model.compile(loss=loss, optimizer=optimizer)
+            model.summary()
+
+            # Fit the model to the data
+            history = model.fit(features, labels, epochs=epochs, validation_split=test_split)
+            hist = pd.DataFrame(history.history)
+            hist["epoch"] = history.epoch
+            info(hist)
+
+            last_row = hist.loc[hist["epoch"] == epochs - 1].values[0]
```
Comment (NFabrizio) [inline modelservice/worker.py:160]: I'm not super familiar with pandas. How does passing a single boolean value to `.loc` work? I thought you had to pass an array with length equal to the number of rows when using boolean values.

File: modelservice/worker.py [hunk 3]
```diff
@@ -72,12 +78,118 @@ def register(self):
...
+                name: str = hidden_layer.name
+                units: int = hidden_layer.units
+                activation = "relu"
+                model.add(tf.keras.layers.Dense(units=units, activation=activation, name=name))
+
+            # Add output layer
+            output_layer: OutputLayer = hyper_parameters.output_layer
+            name: str = output_layer.name
+            activation = "relu"
+            model.add(tf.keras.layers.Dense(units=1, activation=activation, name=name))
```
Comment (NFabrizio) [inline modelservice/worker.py:148]: 👍  for adding hidden and output layer data in the hyperparameters and for keeping it simple and just using relu.

File: modelservice/worker.py [hunk 4]
```diff
@@ -72,12 +78,118 @@ def register(self):
...
+                training_loss=training_loss,
+                validation_loss=validation_loss,
+                duration_sec=gis_join_timer.elapsed,
+                error_occurred=False,
+                error_message="",
+                gis_join_metadata=GisJoinMetadata(
+                    gis_join=gis_join,
+                    count=len_df
+                )
+            )
```
Comment (NFabrizio) [inline modelservice/worker.py:176]: In our proposal, we also mentioned measuring benchmarks for disk I/O, network I/O, page fault rates, memory usage and CPU usage for the duration of each job. Do we want to include that in our metric object, as a separate metric or do we want to skip these benchmarks for this project?

File: modelservice/worker_test.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import pandas as pd
+from pprint import pprint
+
+import logging
+from logging import info, error
+from sklearn.preprocessing import MinMaxScaler
+
+
+# Strictly for testing
+def main():
```
Comment (NFabrizio) [inline modelservice/worker_test.py:12]: Not at all critical, but would it be simpler to have this test call the BuildModels method of worker.py so that we always have the most recent changes from there?

File: experiments/README.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# Experiments
+
+- RQ-1 How can we scale training/validating of geospatial models?
+  - Train models in reasonable amount of time
+- RQ-2 How can we parameterize the configuration of the model?
+  - Parameterizing models allows for users to specify predefined hyperparameters which saves time
+- RQ-3 How can we efficiently store and retrieve trained models?
```
Comment (NFabrizio) [inline experiments/README.md:7]: Minor comment: I don't think we are planning to test anything around how we store the models, but just the retrieval times, so we can probably just focus on retrieval in this question.

File: experiments/README.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+## RQ-2
+
+
+
+## RQ-3
+
+
+## RQ-4
+
```
Comment (NFabrizio) [inline experiments/README.md]:
```suggestion
Create variogram of global versus local model losses.
```

File: experiments/README.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+- RQ-3 How can we efficiently store and retrieve trained models?
+- RQ-4 How do models trained locally compare to a globally-trained model?
+
+## RQ-1
+
+Compare time of global model to time of all local models.
+
+
+## RQ-2
+
```
Comment (NFabrizio) [inline experiments/README.md:16]:
```suggestion
N/A - This is just a statement of logical fact.
```

File: experiments/README.md (hunk 6 above)
Comment (NFabrizio) [reply experiments/README.md:7]: Need to determine the best way to build this experiment (e.g., random sample of models to retrieve, base it on model size and retrieval time, etc.). Since we are only planning to set up retrieval in one way, will measuring this add much value to our project?

**Comment by NFabrizio** [comment]
Just a few minor comments and questions. Overall, this looks great.

**Review by NFabrizio — APPROVED** [review]

File: modelservice/worker.py (hunk 1 above)
Comment (inf0rmatiker) [reply modelservice/worker.py:190]: Maybe if we were putting this into production ;)

File: modelservice/worker.py (hunk 2 above)
Comment (inf0rmatiker) [reply modelservice/worker.py:160]: This works as a filter, but you're right we should be using this: https://stackoverflow.com/questions/70811348/python-get-last-row-from-pandas-data-frame

File: modelservice/worker.py (hunk 4 above)
Comment (inf0rmatiker) [reply modelservice/worker.py:176]: Can't get those from python runtime, has to be out of band

File: modelservice/worker_test.py (hunk 5 above)
Comment (inf0rmatiker) [reply modelservice/worker_test.py:12]: No because of the way Python packages work, if you try to `from modelservice import ...` and then run this via `python3.8 worker_test.py` it won't know what the `modelservice` package is. In order to be able to use absolute package imports like that, you have to run it from project root like `python3.8 -m modelservice <args>` and it automatically executes `__main__.py`, so there's not a good way to execute this specific `main()` function in this test file

File: experiments/README.md (hunk 6 above)
Comment (inf0rmatiker) [reply experiments/README.md:7]: We can just retrieve N models and measure network I/O for the machines
